src/ray/raylet/actor_registration.h

```diff
@@ -118,8 +132,28 @@ class ActorRegistration {
   /// dependencies. This is indexed by handle.
   std::unordered_map<ActorHandleID, FrontierLeaf> frontier_;
 
-  /// All of the dummy object IDs from this actor's tasks.
-  std::vector<ObjectID> dummy_objects_;
+  /// A map from all of the dummy object IDs stored for this actor to the
+  /// number of actor handles that have the object ID in scope. Each actor
+  /// handle will have one dummy object ID in scope, the execution dependency
+  /// for the next task submitted by that actor handle to be executed. If a
+  /// task submitted by this actor handle has been executed before, then this
+  /// is also the dummy object returned by the last task executed. Otherwise,
+  /// it's the dummy object returned by the last task submitted on the original
+  /// actor handle from which this handle was forked.
+  ///
+  /// We release an actor handle's dummy object when a new task submitted by
+  /// that handle finishes, since this means that the handle will never submit
+  /// another task that depends on that object. We also acquire a reference to
+  /// the dummy object returned by the finished task, since this means that the
+  /// next task submitted by the handle will depend on that object. If new
+  /// actor handles are created, these will be included in the next task
+  /// submitted on the original actor handle and are given a reference to the
+  /// same dummy object as the original handle.
+  ///
+  /// Once all handles have released a dummy object, it will be removed from
+  /// this map. This object is safe to evict, since no handle will submit
+  /// another method dependent on that object.
+  std::unordered_map<ObjectID, int64_t> dummy_objects_;
```

Comment:
It took me quiet a while to understand the above comment.
Base on my understanding, I'm trying to summarize the comment and make it easier to understand for future readers. 

-------

This map is used to track all the unreleased dummy objects for this actor. The map key is the dummy object ID, and the map value is the number of actor handles that depends on this dummy object. When the map value decreases to 0, the dummy object is safe to release from the object manager.

An actor handle depends on a dummy object when its next unfinished task depends on the dummy object. For a given dummy object (say D), there could be 2 types of such actor handles:
1) The actor handle (say H) on which D's creating task (say T) was submitted. If T's next task hasn't finished yet, H still depends on D.
2) Any handles that were forked from H after T finished, and before T's next task finishes.  Such handles depend on H until their first tasks finish.

Reply:
Thanks! I agree that this is clearer.

Reply:
I can not understand this very well. Because all calls in an Actor form a line in timeline, then should not all the handles for the same Actor depends on the latest same dummy object instead of each handle's last task dummy object?

Reply:
Hmm I'm not sure exactly what you're asking, but the dummy objects represent the last task that was known to execute on an actor. So when a task is submitted, it depends on the last task submitted on the same handle. When a task is executed, we might update that dependency to reflect the last task that was actually executed on the actor, which may have come from a different handle.

Reply:
Then I mean why should we keep each dependency for all ActorHandles of the same Actor instead of just keeping the latest one? Because we will always update the ActorHandle dependency  to the latest one when execute the actor task, then what is the sense to keep a dependency for each ActorHandle when submit the task?

Reply:
I think I understand your question now. We keep the dependency for each ActorHandle so that we know which tasks have executed so far and which task from each handle can execute next. The former is mostly important for failure scenarios (e.g., making sure you don't re-execute the same task twice on an actor), and the latter comes up during normal execution when there are multiple tasks from the same handle that could execute.

Both of these things don't strictly need the "dependency"; you could accomplish the same thing with task counters for each handle. However, it's convenient to use dependencies since this is also how we determine whether non-actor tasks can be executed.

Reply:
That is to say, when re-execute an Actor Method Sequences after failure, we can only guarantee call order in one ActorHandle, but not in global order(across all handles) with the last Actor method sequences?